Mathlib/NumberTheory/Cyclotomic/PID.lean

```diff
@@ -0,0 +1,58 @@
+/-
+Copyright (c) 2024 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import Mathlib.NumberTheory.NumberField.ClassNumber
+import Mathlib.NumberTheory.Cyclotomic.Rat
+import Mathlib.NumberTheory.Cyclotomic.Embeddings
+
+/-!
+# Cyclotomic fields that are PID.
+We prove that `ℤ [ζₚ]` is a PID for specific values of `p`. The result holds for `p ≤ 19`,
+but the proof is more and more involved.
+
+## Main results
+* `three_pid`: If `IsCyclotomicExtension {3} ℚ K` then `𝓞 K` is a principal ideal domain.
+* `five_pid`: If `IsCyclotomicExtension {5} ℚ K` then `𝓞 K` is a principal ideal domain.
+-/
+
+universe u
+
+namespace IsCyclotomicExtension.Rat
+
+open NumberField Polynomial InfinitePlace Nat Real cyclotomic
+
+variable (K : Type u) [Field K] [NumberField K]
+
+/-- If `IsCyclotomicExtension {3} ℚ K` then `𝓞 K` is a principal ideal domain. -/
+theorem three_pid [IsCyclotomicExtension {3} ℚ K] : IsPrincipalIdealRing (𝓞 K) := by
+  have hp : Fact (Nat.Prime ((3 : ℕ+) : ℕ)) := ⟨Nat.prime_three⟩
```

Comment:
I hate these gymnastics...

Reply:
I don't see how we can avoid it. Using `Fact` works quite well in the general case, but here I don't think we can do something better.

Reply:
I think the root cause of this is the decision to use `PNat`s in the cyclotomic part of the library.
In any case, it may make sense to add `Fact (Nat.Prime (p : PNat))` instances at least for small explicit `p` (and perhaps also something like `instance {p : Nat} [hp: Fact p.Prime] : Fact [Nat.Prime (Nat.toPNat p)] := ...`.)

Reply:
I don't think we use `Fact` in global instance. For example [Nat.prime_three](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Prime.html#Nat.prime_three) is a regular theorem.

Reply:
```lean
instance : Fact (Nat.Prime 3) := inferInstance
```
works.

Reply:
[Nat.fact_prime_three](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Prime.html#Nat.fact_prime_three) (and similar for `2`).

Reply:
Ops, I've missed that!

Reply:
I've added those (also for `5`). It's a little unfortunate that we don't current use [PNat.Prime](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/PNat/Prime.html#PNat.Prime) (what we use is of course defeq, but we write the definition by hand). I will maybe refactor this, but in another PR.

Reply:
And of course we can try to remove `PNat` from the cyclotomic stuff, but that is going to be a lot of work I think.